### PR TITLE
Time Monotonicity Enforcement

### PR DESCRIPTION
**Pull request by AdityaSripal**
## Description

- [x] enforce time monotoncity
- [x] create time misbehaviour
- [x] early return on identical updates
- [x] misbehaviour detection within updateclient

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [inline modules/core/02-client/keeper/client.go:95]: There was too much confusion regarding separation of responsibilities for detecting misbehaviour here. Because conflicting header can be detected here, but time monotonicity can't. Thus, it makes more sense to just make it the responsibility of client developers to do this correctly so we have clear separation of responsibility.

Here i just check if new clientstate is frozen and if so emit appropriate events/write state

I think resulting code is cleaner

**Review by AdityaSripal — COMMENTED** [review]
I've decided to keep the `FrozenHeight` to being the height of the offending header that caused misbehaviour. With time misbehaviour i use the larger height. Since we no longer use its specific value, it doesn't really matter so long as its non-zero. @colin-axner argues for just making the height {0-1} since it isn't used for anything other than a boolean value.

I figure if we're keeping it, we may as well make it meaningful. Care to weigh in @cwgoes ?

File: modules/core/02-client/keeper/client_test.go
Comment (colin-axner) [inline modules/core/02-client/keeper/client_test.go:71]: I refactored this test in #140 to use the new testing package since consensus states were not being stored for all client states resulting status to be `Unknown` on valid cases. 

Could you copy over those changes and adjust?

File: modules/core/02-client/keeper/client_test.go
Comment (colin-axner) [inline modules/core/02-client/keeper/client_test.go:149]: This test case be reduced once you copy over the changes:
```go
{"misbehaviour", func() {
            // perform update and then replicate update
            clientState := path.EndpointA.GetClientState()
            trustedHeight := clientState.GetLatestHeight().(types.Height)
            path.EndpointA.UpdateClient()
            clientState = path.EndpointA.GetClientState()
            fillHeight := clientState.GetLatestHeight().(types.Height)
            updateHeader = createPastUpdateFn(fillHeight, trustedHeight)
        }, false},
        ```

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/misbehaviour_handle.go:36]: we can remove this. This is checked in 02-client

File: modules/light-clients/07-tendermint/types/misbehaviour_handle_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/misbehaviour_handle_test.go:76]: can you add a test where header 1 time is < header 2

File: modules/light-clients/07-tendermint/types/misbehaviour_handle_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/misbehaviour_handle_test.go:372]: can you adjust line 400 to use `suite.chainA.App.AppCodec()`. We should be consistent in either using the old or new testing functions (preferably the new so it is easier to maintain)

File: modules/light-clients/07-tendermint/types/store.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/store.go:199]: can you update the godoc?

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/update_test.go:109]: what is this testing?

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/update_test.go:148]: it's hard to verify this test. There appears to be 3 times:
- client time
- header time
- suite.now

I'm not sure what is before or after without looking into the helper file. Can you add comments as to why this test functions correctly?

I'd assume header time to be after client time. The header is made with the header time, the consensus state is made with the client time, and the previous consensus state is made with header time and the context is set to suite.now. What is the consensus state used for?

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/update_test.go:156]: ditto

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/update_test.go:111]: It worries me a little to use so much manual setup as opposed to just running `SetupClients`, but I guess updating these tests can be done another day

**Review by colin-axner — COMMENTED** [review]
Overall LGTM, nice work!

**Comment by colin-axner** [comment]
changelog entry is needed (I keep forgetting as well)

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:95]: > I don't understand why we cache the context if it is now the full responsibility of the client developers to handle all instances of misbehaviour correctly

They are responsible for telling core IBC if an update was misbehaviour. They are not responsible for rolling back all state changes.

> What if a client wanted to write metadata everytime it handled misbehaviour in an update client message?

This is definitely possible, i guess its up to us what we want to enable. The downside is accidentally leaving in metadata writes after misbehaviour that we intend to write only for valid updates. Think @cwgoes can weigh in on tradeoff between flexibility and opinionated code. I believe being opinionated here and having the ClientKeeper write metadata on valid update makes more sense. Light client implementations are fully responsible for doing update logic (UpdateClient will do none of that).
But ClientKeeper will take the returned output, and do all of the necessary store writes. I think that's a clean separation of responsibility.

> I'd actually prefer to be as defensive as possible. In which case, we should keep the cached context and return an error if a duplicate update occurs without the client detecting misbehavior

I tried doing this and the code got super ugly because there was freezing logic in both the ClientKeeper and in tendermint's update function. That could have been much cleaner if it was just in one place.

Furthermore, I think it's possible to take all client developer checks that must be done by every light client and put them in the ClientKeeper to minimize the possibility of light-client developer error.
But I think in practice, this would make things less secure if it trades off too much on separation of concerns.
Critically, I think it just needs to be clear to a reviewer/developer where a particular check is supposed to happen.
My proposal is that we create a very clear separation of concern that acts as a contract between core IBC and light client developer.

Light client implementation must give core IBC the updated clientstate/consensus state. And it **must** return a frozen client state if the update was evidence of misbehaviour.
Core IBC will in turn store the clientstate (and consensus if valid update), write all callback state changes on successful updates, and emit appropriate events.

This means that there may be redundant checks happening in light clients, that may be missed by some of them. But it gives a very clear rule for what a light client implementation is responsible for. Even though i place responsibility of all misbehaviour checks on light client. As a reader and reviewer I can analyze the light-client implementation **in isolation** and check that it is catching all misbehaviour and holding up its side of bargain.

Without this, I need to be checking whether ClientKeeper+misbehaviour together are catching all misbehaviour. And I need to make sure together they don't miss a gap between them. And that they aren't doing redundant checks. It's also harder as time goes on to determine where a check should go. We would need to make a subjective decision on whether we think some check is universal or not.

For these reasons I think clear separation of concerns is more important than putting all universal checks (even subtle ones) in the ClientKeeper. But yes, this should absolutely be documented in `light_client.md`. Will do so once there's consensus on this point

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:203]: My concern is if misbehaviour is passed in from another module rather than through the SDK msg that we designed

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/misbehaviour_handle.go:96]: The frozen height is the height of the offending header. In the case of time misbehaviour, it is the existence of the larger height that breaks monotonicity.

This feeds into the intuition we had with submitting misbehaviour that lowers frozen height. That logic rested on the understanding that frozen height was height of offending header. You could lower frozen height by submitting proof of an earlier offending header.

No longer matters now, but if we choose to re-enable this would follow that intuition.

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:95]: Regardless of if we allow metadata writes on misbehaviour, we still want to cache so we can discard on error.

Developers shouldn't be forced to revert state themselves on error

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:203]: Out of scope for this PR, let's open separate issue to discuss if there's disagreement

Comment (colin-axner) [reply modules/core/02-client/keeper/client.go:203]: > My concern is if misbehaviour is passed in from another module rather than through the SDK msg that we designed

This can be applied to all our handlers. Is there a use case in mind? We could turn these functions into message servers such that we aren't exposing functions that don't take messages

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/update_test.go:109]: It's testing that submitting a header for which the consensus state already exists passes (no-op)

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (colin-axner) [reply modules/light-clients/07-tendermint/types/misbehaviour_handle.go:96]: Couldn't an attacker construct a fill in header causing the larger height to become the offending header despite it being signed by > 66% of validators and thus steal funds since the frozen height would have been at the larger height

File: modules/light-clients/07-tendermint/types/update_test.go
Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/update_test.go]: The consensus state with header time is just used to create the time violation. I create a monotonic time violation by setting a consensus state at a different height to be the same time as header.

I still need a valid consensus state in order to even pass the update. Note: the trusted height of header points to the valid consensus state at client time.

Documented to make things clearer.

Agreed that client time and header time can be confusing, but it's not just in this test case but throughout tendermint package. future testing pr should resolve this

File: modules/core/02-client/keeper/client.go
Comment (colin-axner) [reply modules/core/02-client/keeper/client.go:99]: I think you make great points. 

> it must return a frozen client state if the update was evidence of misbehaviour.

I agree with this.

> As a reader and reviewer I can analyze the light-client implementation in isolation and check that it is catching all misbehaviour and holding up its side of bargain.

I like this, and I think we can still achieve this with a duplicate check. My concern is that allowing a duplicate update at an existing height is a critical security vulnerability and I'm hesitant to let it go by when we have the capacity to do the check. This is the code I have in mind:

```go
consState, exists := keeper.GetConsensusState()

newClientState, newConsensusState, err := CheckHeaderAndUpdateState()
if err != nil {
    return err
}

// write client state, errors returned later revert state changes
switch {
case: newCilentState.IsFrozen()
    // use logic you have
case: exists && !reflect.DeepEqual(consState, newConsensusState)
    // return err, light client implementation missed misbehaviour handling
default:
    // regular update code
}
```

I don't see why this code gets ugly? It allows light clients to fully implement misbehaviour logic without relying on 02-client and it allows 02-client to prevent duplicate updates which are misbehaviour

File: modules/core/02-client/keeper/client.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:219]: Suppose with Interchain Staking I get a packet of Evidence which is chain-wide misbehaviour. I could freeze the underlying client during processing

File: modules/core/02-client/keeper/client.go
Comment (colin-axner) [reply modules/core/02-client/keeper/client.go:99]: > Regardless of if we allow metadata writes on misbehaviour, we still want to cache so we can discard on error.
> 
> Developers shouldn't be forced to revert state themselves on error

Do you have the use case in mind that update is being called by an external module? Messages that result in errors always have state changes reverted

Comment (AdityaSripal) [reply modules/core/02-client/keeper/client.go:99]: > Do you have the use case in mind that update is being called by an external module? Messages that result in errors always have state changes reverted by baseapp. I think this is a safe assumption to make

Oh yes you're correct about this. We should only cache if we discard metadata on misbehaviour

File: modules/core/02-client/keeper/client_test.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client_test.go:71]: I don't have this function: `suite.chainA.ConstructUpdateTMClientHeaderWithTrustedHeight`

File: modules/light-clients/07-tendermint/types/misbehaviour.go
Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/misbehaviour.go:39]: Out of scope, let's create an issue and discuss

Comment (colin-axner) [reply modules/light-clients/07-tendermint/types/misbehaviour.go:39]: see #148

File: modules/core/02-client/keeper/client.go
Comment (colin-axner) [reply modules/core/02-client/keeper/client.go:219]: see #150

File: modules/core/02-client/keeper/client_test.go
Comment (colin-axner) [inline modules/core/02-client/keeper/client_test.go]: I think in the future it is best to avoid setting all the consensus states manually, but this is fine. I think setting the update headers consensus state is fine, but the previous/intermediate consensus state (setup code) should just be done by calling `path.EndpointA.UpdateClient`. I think it helps in maintainability/readability in testing code

In my opinion, hard coding the heights is generally harder to maintain as opposed to referencing the latest height before/after an update.

File: modules/core/02-client/keeper/client_test.go
Comment (colin-axner) [inline modules/core/02-client/keeper/client_test.go:164]: `suite.header.Header.Time.Add(2 * time.Hour)` should be `suite.coordinator.CurrentTime.Add(2 * time.Hour)`

**Review by colin-axner — COMMENTED** [review]
LGTM, waiting to approve once we resolve decisions on
- using cache ctx (no preference)
- returning error if client did not handle misbehaviour (duplicate update produces different consensus state)

I have a preference for doing the consensus state check, but willing to concede if others consider it unnecessary

**Comment by cwgoes** [comment]
> I figure if we're keeping it, we may as well make it meaningful. Care to weigh in @cwgoes ?

Just to refresh my memory, the goal here is backwards compatibility, right? Otherwise it seems odd to me to keep a field with a range (integer) far greater than the set of semantically meaningful values (0, non-zero). For backwards compatibility I guess there isn't too much difference between 0, 1 and 0, nonzero height - although the latter does run the risk of appearing to be more semantically relevant than it actually is - and we don't really want IBC users to start using this value for other kinds of data processing, if they really care about misbehaviour heights they should read them from the event logs - those are both some (weak) reasons to prefer 0, 1 I think.

File: modules/core/02-client/keeper/client_test.go
Comment (AdityaSripal) [reply modules/core/02-client/keeper/client_test.go]: These are all past updates so it should be ok

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/types/misbehaviour_handle.go:92]: can we have a const `FrozenHeight`? This would be useful in tests

Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/misbehaviour_handle.go:92]: Struct can't be a constant but I can make it a var inside tendermint package

File: modules/core/02-client/keeper/client.go
Comment (cwgoes) [reply modules/core/02-client/keeper/client.go:98]: > My concern is that allowing a duplicate update at an existing height is a critical security vulnerability and I'm hesitant to let it go by when we have the capacity to do the check.

Here's a question - is this always true? Certainly it is a problem if unequal consensus states at the same height would allow for violation of exactly-once packet delivery guarantees or timeouts, but there could conceivably be client types which _allow_ duplicate consensus states, just not verification at them (so they are only intermediate update points) - for example, a (non-Tendermint) consensus algorithm could have a block history which looks like this:

blocks

Is this a case we want to consider? There is something to be said for not overly constraining what it means for clients to be "correct", since clients implement all of the packet data / timeout / etc. verification functions anyways.

Comment (colin-axner) [reply modules/core/02-client/keeper/client.go:98]: Great question! I didn't realize intermediate update points were a possibility. 

In light of our [discussion yesterday](https://github.com/cosmos/ibc/issues/569#issuecomment-832051686), I don't see the usefulness of adding this check if in the near future, light client implementations will be responsible for getting/setting client/consensus states. In this design, light clients should definitely be aware to guard against duplicate updates which constitute misbehaviour

**Review by colin-axner — APPROVED** [review]
Excellent work! In light of our [discussion yesterday](https://github.com/cosmos/ibc/issues/569#issuecomment-832051686), I think the current solution aligns perfectly with future changes

**Comment by colin-axner** [comment]
@AdityaSripal I fixed the merge conflicts. Everything was straightforward except for the check in update client for if the client is frozen. I update the code to be semantically equivalent:

Before
```go
if !client.IsFrozen {

} else {

}
```

after
```go
if status := client.Status; status != exported.Frozen {

} else {

}
```

It is just slightly odd because we aren't checking if the client is active, just that it is not frozen

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (cwgoes) [inline modules/light-clients/07-tendermint/types/misbehaviour_handle.go:54]: just to confirm, do we check this somewhere?

**Review by cwgoes — APPROVED** [review]
ACK modulo quick confirmation question

File: modules/light-clients/07-tendermint/types/misbehaviour_handle.go
Comment (AdityaSripal) [reply modules/light-clients/07-tendermint/types/misbehaviour_handle.go:54]: We check it on `ValidateBasic` of Misbehaviour

**Pull request merged**
